### DISCUSSION TITLE
API request within extension: Connection refused

### DISCUSSION THREAD

**Discussion by GitYepper** [ckan/ckan#6315] [Q&A]
I just wanted to make some test API requests (like https://docs.ckan.org/en/2.9/api/index.html#making-an-api-request or https://docs.ckan.org/en/2.9/api/index.html#example-importing-datasets-with-the-ckan-api) within my CKAN extension but I always get an error that the connection was refused when using the urllib2 library. 

`urllib2.URLError: <urlopen error [Errno 111] Connection refused>`

What makes it a bit confusing for me: When I run the correpsonding commands on the command line with curl (for example to create a package or a dataset) everthing works without any errors, so the URL seems to be correct/accessible. The URL by the way is the same as `ckan.site_url`.

To summarize it:

This works (via command line):
` curl -X POST http://<SERVER_IP>:5000/api/3/action/package_create` ....

This leads to the error from above when calling `urllib2.urlopen` (code line in my extension):
`urllib2.Request("http://<SERVER_IP>:5000/api/3/action/package_create")`

Could it be a problem that I want to call the CKAN API within my CKAN extension? The authorization/API key should be fine..

Would be thankful for any feedback/input why I cannot access the API within my extension.

**Comment by jqnatividad** [comment]
If you look at the [Writing Extensions tutorial](https://docs.ckan.org/en/2.9/extensions/tutorial.html), the best practice is to use the plugin toolkit so you can pass the authorization context.

Alternatively, I've found it easier to use the [ckanapi python module](https://github.com/ckan/ckanapi#ckanapi-python-module).

Also, please close #6316, as Discussions is the right place to post questions and the Issue Tracker is for bug reports.

**Comment by GitYepper** [comment]
Thank you for your advice. 

At least I was now able to create new packages via the toolkit.get_action() function but couldn't create new resources.

So this is working now:
```
data_dict = {
        'name': 'test-package',
        'title': 'Test Package',
        'private': False,
        'owner_org': 'XXXX',
    }
   
    context = {'user': 'XXXX'}
    toolkit.get_action('package_create')(context, data_dict)
```

While this leads to the following error:
```
    data_dict = {
        'package_id': 'test-package',
        'url': 'dummy'
    }
   
    context = {'user': 'XXXX'}
    toolkit.get_action('resource_create')(context, data_dict)
```

` File "/usr/lib/ckan/default/lib/python2.7/site-packages/flask/helpers.py", line 307, in url_for
    "Attempted to generate a URL without the application context being"
RuntimeError: Attempted to generate a URL without the application context being pushed. This has to be executed when application context is available.
`

Any idea how to fix this?



----------------------------------------

Btw, also with the ckanapi python module I cannot create resources - using RemoteCKAN leads to the exact same connection error like in the original post

**Comment by jqnatividad** [comment]
Can you try adding `name` to the data_dict?

Also, any particular reason you're using Python 2.7 now that its EOL and CKAN 2.9 supports Python 3?

As for ckanapi, have you tried using this particular snippet against your installation?
https://github.com/ckan/ckanapi#file-uploads

**Reply by GitYepper** [reply]
Unfortunately, I get the same error with the application context when adding `name`  to the data_dict.

The snippet from ckanapi leads to a refused connection:
`requests.exceptions.ConnectionError: HTTPConnectionPool(host='<SERVER_IP>', port=5000): Max retries exceeded with url: /api/action/resource_create (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f2a229ca350>: Failed to establish a new connection: [Errno 111] Connection refused',))`

As <SERVER_IP> I used the IP for accessing my CKAN instance via the browser (same as `ckan.site_url`).

No, I don't use Python 2.7 for any particular reason.. I will update it soon. (Already had it updated but it broke solr, so I switched back for now as I got the same errors with Python 3 anyway)

**Comment by GitYepper** [comment]
I am closing this discussion as I chose another approach without creating an own extension. 

@jqnatividad Thank you anyway, the reference to the ckanapi module was quite useful and is working for me now (outside my extension code).

EDIT: Ok, just saw I cannot close a discussion, but just wanted to say that I don't need help on this issue anymore ;)
